File: 06/JYC/28.ts
function strStr(haystack: string, needle: string): number {
  function getPartialMatch(needle: string): number[] {
    const m = needle.length;
    const pi = new Array(m).fill(0);
    let begin = 1;
    let matched = 0;
    while (begin + matched < m) {
      if (needle[begin + matched] === needle[matched]) {
        matched++;
        pi[begin + matched - 1] = matched;
      } else {
        if (matched === 0) {
          begin++;
        } else {
          begin += matched - pi[matched - 1];
          matched = pi[matched - 1];
        }
      }
    }
    return pi;
  }
  const n = haystack.length;
  const m = needle.length;
  const pi = getPartialMatch(needle);
  let begin = 0;
  let matched = 0;
  while (begin <= n - m) {
    if (matched < m && haystack[begin + matched] === needle[matched]) {
      matched++;
      if (matched === m) {
        return begin;
      }
    } else {
      if (matched === 0) {
        begin++;
      } else {
        begin += matched - pi[matched - 1];
        matched = pi[matched - 1];
      }
    }
  }
  return -1;
}

File: 06/ahnjaewoo/28.cpp
class Solution {
public:
    vector<int> getKmp(string needle) {
        int length = needle.size();
        int start = 1;
        int match = 0;

        vector<int> p(length, 0);

        while (start + match < length) {
            if (needle[start + match] == needle[match]) {
                match += 1;
                p[start + match - 1] = match;
            }
            else {
                if (match == 0) start += 1;
                else {
                    start += match - p[match - 1];
                    match = p[match - 1];
                }
            }
        }
        return p;
    }

    int strStr(string haystack, string needle) {
        int h_len = haystack.size();
        int n_len = needle.size();
        int start = 0;
        int match = 0;
        vector<int> p = getKmp(needle);
 
        while (start + n_len <= h_len) {
            if (match < n_len && haystack[start + match] == needle[match]) {
                match += 1;
                if (match == n_len) {
                    return start;
                }
            }
            else {
                if (match == 0) start += 1;
                else {
                    start += match - p[match - 1];
                    match = p[match - 1];
                }
            }
        }
        return -1;
    }
};
